Minimise articles

lc-thomasberger opened the pull request:
Fixes #1483

taylortom commented:
Think we should push this one to the next release for testing reasons

lc-thomasberger commented:
@taylortom fine for me. Maybe with some more minor UI enhancements?

canstudios-louisem commented on frontend/src/core/models/articleModel.js:
```diff
@@ -6,7 +6,12 @@ define(function(require) {
+    _children: 'blocks',
+
+    defaults: {
+      _isCollapsed: false
```
Can you confirm this is not saved to the DB and therefore is exported on build?

lc-thomasberger replied:
It is not saved in the DB and not exported. This is the article.json from an exported course:
```
{
    "_id": "5a7d5be53fbfedfc836c023d",
    "_type": "article",
    "_parentId": "5a7d5be53fbfedfc836c023c",
    "_onScreen": {
      "_isEnabled": false,
      "_classes": "",
      "_percentInviewVertical": 50
    },
    "_isVisible": true,
    "_isHidden": false,
    "_isAvailable": true,
    "_isOptional": false,
    "instruction": "",
    "_classes": "",
    "body": "",
    "displayTitle": "collapsed article ",
    "title": "collapsed article "
  }
```

canstudios-louisem commented:
Are all blocked collapsed by default? if so could we have a button that expands and collapses them all? It would make it a bit easier if you are searching for a specific thing.

lc-thomasberger commented:
@canstudios-louisem Default is not collapsed. When a user collapses an article, it only sets the height of the article and uses overflow hidden to hide the blocks.
The "collapsing" is only applied as long as you work in a course and not synced with other users. If you switch pages inside one course, the settings for collapsed is saved. If you leave the course, the settings are lost.

dancgray commented:
Nice feature, if the collapse toggle was a button outside the context menu it would halve the number of clicks required.

taylortom commented:
I agree with dan -- for usability, I think we should look at perhaps adding a specific minimise button

lc-thomasberger commented:
That's true @dancgray and @taylortom . Contextmenu will be pretty "crowded". Will add a button.

lc-thomasberger commented:
When I added the button I had to update the "toolbar". I've used flexbox for this and was surprised how well it worked in IE. This refactor should make future amends easier. 

![image](https://user-images.githubusercontent.com/11990374/36531381-ee5f9a1c-17bd-11e8-963b-5562bc4346c6.png)

lc-thomasberger commented:
Please note that I update options module to reset the selected element. `Origin.trigger('options:reset:ui', 'optionGroup');` I should have made a single PR for this ...

taylortom commented:
Wasn't convinced about this initially, but works well, and think people will find it really useful. A few suggestions:
- I'm not sure about the icons; these are usually used for full-screen/non-full-screen. What do you think of either a [plus](https://fontawesome.com/icons/plus?style=solid)/[minus](https://fontawesome.com/icons/window-minimize?style=solid) or [window-maximise](https://fontawesome.com/icons/window-maximize?style=regular)/minus?
- Can we change the tooltip on hover to show expand/collapse depending on the state? (it just shows collapse currently)
- I'm not sure about the collapsed styling (specifically being able to see some of the inner block). My initial thought is to just show the title bar -- I'm willing to be persuaded on this one, as I see arguments for both!

lc-thomasberger commented:
Thanks @taylortom 
I think window-maximise is too complicated. Thought about it initially but didn't like it. I also tried a [arrow](https://fontawesome.com/icons/angle-down?style=solid) and rotate it to indicate the states. Plus / Minus would make sense as it is used in Accordion.

Options would need some improvements. I think we should use flexbox to align the items. This will make layouts easier and we should be able to add a title to groups.

Tooltip based on state should be fine.

I get your point about the last one. I wanted to show that the blocks are still there and not "deleted". I may play with some styles and post screenshots in gitter.

taylortom commented:
Just noticed github have very similar functionality in their diff stuff (click the 'Files changed' tab up at the top of this page). They use up/down chevrons, and only show the title bar.

dancgray commented:
Works well for me, but I can imagine people will want the minimised articles preserved after editing an article/block/component on the screen.

lc-thomasberger commented:
@dancgray yes, there was an update with the editor data. The models are now refetched whenever something is edited. We could store the toggled state in the db, which will apply the selection to every other user or use a model to store the toggle state for each element and apply it once they are rendered.

For now, I added the global toggle buttons at the top. This makes it easier to collapse all articles at once.

taylortom commented:
What about storing the prefs in BrowserStorage like https://github.com/adaptlearning/adapt_authoring/pull/1378 ?

lc-thomasberger commented:
@taylortom this was sort of the idea with the model approach. Not sure if it should persist sessions, as this will likely spam your local storage and cleaning up will be hard. Hard to check if a article still exists if it is not deleted by the user.

lc-thomasberger commented:
Restore collapsed state and use plus / minus icons:
![image](https://user-images.githubusercontent.com/11990374/37587215-f9847340-2b5e-11e8-9e53-fd0f3b471168.png)

Have not updated the options view. This will need a proper update in a new PR.

canstudios-louisem commented:
Its a tiny thing but can we put these onto one line all icons in the AT are in line with their text. It should also probably say `Collapse Articles` `Expand Articles` so its a bit more clear what they do.

![image](https://user-images.githubusercontent.com/26000433/37780449-b0740a5a-2de6-11e8-93dc-7675b6315837.png)

lc-thomasberger commented:
@canstudios-louisem alignment was fixed in this PR: https://github.com/adaptlearning/adapt_authoring/pull/1925

I updated the whole options layout to properly align the items using flexbox.

canstudios-louisem commented:
@lc-thomasberger why is that a sperate PR? Seems rather unnecessary to review it then review it again.

lc-thomasberger commented:
I was not sure how much needed changing so I started with a separate branch. I think it can be merged in this one directly so both can be reviewed at once.

taylortom commented:
Just found the screenshots I did for this:

Using the existing icons: 
![screen shot 2018-03-23 at 11 05 12](https://user-images.githubusercontent.com/1059083/38421520-4d044236-399f-11e8-8d5d-6620722ac448.png)

Using the chevrons (my preference):
![screen shot 2018-03-23 at 11 15 43](https://user-images.githubusercontent.com/1059083/38421523-513563bc-399f-11e8-8e3e-350042f042eb.png)

taylortom commented:
Looks good to me, thanks @lc-thomasberger! 

This makes dragging articles around a lot more intuitive. Once this PR has gone in, I wonder if we should look into minimising all articles on drag by default...

One minor thing -- I think the icons at the top need switching (i.e. the up chevron for 'collapse all', down for 'expand all')

lc-thomasberger commented:
@taylortom I've replicated githubs logic. down when expanded and up when collapsed.

taylortom commented:
Yeah I think it works on the articles themselves as it acts as a sort of status indicator, but seems the wrong way round in the buttons at the top (happy to go with the general consensus if people think otherwise though).

lc-thomasberger commented:
I think icon state for expand / collapse should always be the same.

taylortom commented:
Yeah that's fair, I think we maybe just need to invert them all then - which do you think is most intuitive?

lc-thomasberger commented:
Tbh. I never get the logic which chevron shows which state. 😄 That's why I decided to go for the plus and minus.
I am absolutely fine to go with chevrons though. It's only important that the icons are consistent for each state. If not it is confusing for the user.

taylortom commented:
Yeah I agree, maybe using [fa-plus](https://fontawesome.com/icons/plus?style=solid) and [fa-minus](https://fontawesome.com/icons/minus?style=solid) is the way to go 😅

lc-thomasberger commented:
I shall give it a go 😄

dancgray commented:
This works really well for me now. Noticed a couple of small things though, the trash icon on the menu editor is different 'purple' colour, and I can no longer double click on a component to edit it.

dancgray commented:
@lc-thomasberger, rebuilt but still cannot double click over the body of the component to edit which you used to be able to do:
![screen shot 2018-04-11 at 13 52 28](https://user-images.githubusercontent.com/6452430/38617837-f11b6b94-3d8f-11e8-9764-49b9307f2280.png)

I can double click when cursor is over component-detail-inner div:
![screen shot 2018-04-11 at 13 51 46](https://user-images.githubusercontent.com/6452430/38618028-67c7aaaa-3d90-11e8-97db-8c1465d62029.png)
Let me know if this is intended.

lc-thomasberger commented:
@dancgray thanks for posting the screenshots, I misunderstood your first query. Initially I updated the dbClick listener to only match the detail div. As @taylortom raised an issue with triggering the dbClick event when toggling article collapsed state. 
I updated the PR so now you can dbClick on the "body" of each element but excluded the buttons in the detail div. 
Also updated the styling of the menuItem's. It uses buttons instead of <a> and flexbox to align the content.

tomgreenfield commented on frontend/src/modules/editor/contentObject/views/editorPageArticleView.js:
```diff
@@ -11,25 +11,37 @@ define(function(require){
+      'click .add-block': 'addBlock',
+      'click .article-delete': 'deleteArticlePrompt',
+      'click .open-context-article': 'openContextMenu',
+      'dblclick .editor-page-element-detail-inner,.article-blocks': 'loadArticleEdit',
```
Please restore the original generic `dblclick` listeners so we may click anywhere to edit e.g. on an article's body text.

The code you've implemented in `editorOriginView.js` should be enough to prevent double clicking on buttons?

tomgreenfield commented on frontend/src/modules/editor/contentObject/less/editorPageArticle.less:
```diff
@@ -20,13 +24,25 @@
+        
+        .article-blocks {
+            transition: opacity 200ms ease-in-out;
+            opacity: 0.5;
```
Is the `opacity` property needed?

tomgreenfield commented on frontend/src/modules/editor/global/less/editor.less:
```diff
@@ -100,35 +100,80 @@
+      width: 100%;
+      box-sizing: border-box;
+      display: -ms-flexbox;
+      display: -webkit-flex;
```
Prefixed attributes not required in modern browsers.

lc-thomasberger replied:
but we can leave them in ... as it will help support older versions and not break modern browsers

taylortom replied:
I'd say we should probably try to avoid adding anything specifically to support legacy browsers (particularly as the framework no longer supports pre-v11 IE).

lc-thomasberger commented on frontend/src/modules/editor/contentObject/templates/editorPageArticle.hbs:
```diff
@@ -1,16 +1,15 @@
+    {{>part_editorItemSidebar}}
   </div>
 
+  {{#if body}}
```
Tried to not touch the dom structure as this might break existing theme amends.

tomgreenfield commented:
`overflow-y: scroll` will work fine with infinite scrolling – it's nearly always same behaviour as `auto`, with the addition of a disabled scrollbar when content doesn't overflow.

Therefore, happy for it to go on `contentPane` or `editor-view`.

tomgreenfield commented:
I do see what you mean about `overflow-y` on the main content pane; the fixed position of Asset's Management preview container causes some whitespace in the layout.

Best to stick to `editor-view` as you suggest.

lc-thomasberger commented:
@canstudios-louisem may be related to markup update, ist this something that bothers us?

canstudios-louisem commented:
@lc-thomasberger i think its something that should be fixed as it doesn't look right now.

taylortom commented:
I've just made a few minor tweaks to the padding to look more consistent in the open/closed state. Most notable being the increased padding at the bottom of the article title when minimised (and increased page title to match). I can push this to **adaptlearning** if we want it (see commit [here](https://github.com/cgkineo/adapt_authoring/commit/c28bdd542943ef47ce06a0dd42666a5515a44716)).

See below for comparison.

### Before
![screen shot 2018-04-25 at 13 07 18](https://user-images.githubusercontent.com/1059083/39244793-ad230ed8-4889-11e8-8f27-cd19022d3758.png)

### After
![screen shot 2018-04-25 at 13 03 12](https://user-images.githubusercontent.com/1059083/39244640-2939a5aa-4889-11e8-8106-cad39d693445.png)

tomgreenfield commented:
Yep, please push @taylortom.

taylortom commented:
Todo:
- Padding on page body
- Initial animation

lc-thomasberger commented:
are we going to fix the initial animation? @taylortom

lc-thomasberger commented:
Have a fix for the padding of page-header, articIe-boy and block-body I am happy to push if we like the design.
before: https://files.gitter.im/lc-thomasberger/8de6/image.png

after: https://files.gitter.im/lc-thomasberger/sRRp/image.png

taylortom commented:
@lc-thomasberger yeah that looks great to me!
